[Impeller] Avoid encoding metal commands while the GPU is unavailable when decoding images.

dnfield opened the pull request:
Fixes https://github.com/flutter/flutter/issues/126878

This disables device private upload on iOS when backgrounded, and disables mipmap generation when backgrounded.

We don't have a good way to test the core problem in this repo because it only reproduces on physical iOS hardware - simulators don't really care if you do this stuff in the background.

I'll write a devicelab test in the framework to capture this. In the mean time it can be reviewed.

We could consider making the IOManager a shared_ptr instead of having an fml::WeakPtr and that'd clean up some of the extra arguments to engine construction - or we could consider vending the sync switch from impeller::Context unconditionally, but it's pretty iOS specific...

gaaclarke commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -370,32 +373,44 @@ ImageDecoderImpeller::UploadTextureToShared(
-      return std::make_pair(nullptr, decode_error);
+    std::optional<std::string> decode_error;
+#if FML_OS_IOS
+    gpu_disabled_switch->Execute(fml::SyncSwitch::Handlers().SetIfFalse(
```
Should the true branch be setting `decode_error`?

gaaclarke replied:
We ripped out opengl for iOS right?

dnfield replied:
Correct

gaaclarke commented on lib/ui/painting/image_decoder.h:
```diff
@@ -27,7 +27,8 @@ class ImageDecoder {
       std::shared_ptr<fml::ConcurrentTaskRunner> concurrent_task_runner,
-      fml::WeakPtr<IOManager> io_manager);
+      fml::WeakPtr<IOManager> io_manager,
+      const std::shared_ptr<fml::SyncSwitch>& gpu_disabled_switch = nullptr);
```
Everyone should have a syncswitch so there's no need to support nullptr.

gaaclarke commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -446,24 +461,39 @@ void ImageDecoderImpeller::Decode(fml::RefPtr<ImageDescriptor> descriptor,
+                [&needs_upload, &image, &decode_error, &context,
+                 &bitmap_result] {
+                  needs_upload = false;
+                  FML_LOG(ERROR) << "Yay";
```
Stray log.

dnfield replied:
Done

dnfield replied:
(removed the one below too)

gaaclarke commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -446,24 +461,39 @@ void ImageDecoderImpeller::Decode(fml::RefPtr<ImageDescriptor> descriptor,
-#endif
+          FML_DCHECK(gpu_disabled_switch);
+          if (gpu_disabled_switch) {
+            gpu_disabled_switch->Execute(fml::SyncSwitch::Handlers().SetIfFalse(
```
It's worth adding a comment about the omission of a true branch.

dnfield replied:
This got moved and there's no longer an omission.

gaaclarke reviewed (changes requested):
Looks good for the most part.  The only thing that should get changed is removing the conditional compilation.  I think we should treat sync_switch as a nonnull parameter instead.  Also, I want clarity in for the true branches of the syncswitches.  We should have comments that at least explain their omission.

dnfield commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -370,32 +373,44 @@ ImageDecoderImpeller::UploadTextureToShared(
-      FML_DLOG(ERROR) << decode_error;
-      return std::make_pair(nullptr, decode_error);
+    std::optional<std::string> decode_error;
+#if FML_OS_IOS
```
Sounds good, that will actually make this more testable too.

dnfield commented:
Added a test.

Ideally this would get a screenshot based test on a real device, but that will probably take some more time to write. For now the test I added validates that if the sync switch is disabled we do not create any command buffers.

gaaclarke commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -318,10 +313,40 @@ ImageDecoderImpeller::UploadTextureToPrivate(
+          .SetIfTrue([&result, context, bitmap, gpu_disabled_switch] {
+            // create_mips is false because we already know the GPU is disabled.
+            result = UploadTextureToShared(context, bitmap, gpu_disabled_switch,
+                                           false);
```
```suggestion
                                           /*create_mips=*/false);
```

gaaclarke commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -318,10 +313,40 @@ ImageDecoderImpeller::UploadTextureToPrivate(
+            result = UnsafeUploadTextureToPrivate(context, buffer, image_info);
+          })
+          .SetIfTrue([&result, context, bitmap, gpu_disabled_switch] {
+            // create_mips is false because we already know the GPU is disabled.
```
Can't this result in suboptimal rendering based on when the gpu is disabled?  Are we going to get bug reports about blurry images that we can't source back to coming from backgrounded apps?

dnfield replied:
It can result in suboptimal rendering. We could choose to build mips when the GPU is available again, but that's going to be a more invasive/risky change. I'd like this to be cherry pickable.

gaaclarke commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -446,24 +479,27 @@ void ImageDecoderImpeller::Decode(fml::RefPtr<ImageDescriptor> descriptor,
-              UploadTextureToShared(context, bitmap_result.sk_bitmap);
-#endif
+          std::tie(image, decode_error) = UploadTextureToShared(
+              context, bitmap_result.sk_bitmap, gpu_disabled_switch, true);
```
```suggestion
              context, bitmap_result.sk_bitmap, gpu_disabled_switch, /*create_mips=*/true);
```

gaaclarke commented on shell/common/engine_unittests.cc:
```diff
@@ -162,7 +162,8 @@ TEST_F(EngineTest, Create) {
         /*font_collection=*/std::make_shared<FontCollection>(),
-        /*runtime_controller=*/std::move(runtime_controller_));
+        /*runtime_controller=*/std::move(runtime_controller_),
+        /*gpu_disabled_switch=*/nullptr);
```
nit: make syncswitches here since technically it isn't supposed to be null.  We could also add a dcheck that it isn't null.

bdero commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -370,32 +395,40 @@ ImageDecoderImpeller::UploadTextureToShared(
+            return;
+          }
+          command_buffer->WaitUntilScheduled();
+        }));
```
This isn't something we should solve now, but we'll eventually need a solution to defer generating mipmaps. Not generating mipmaps will cause min filtered images to continue appearing more blurry than Skia once we start properly setting the mip bias.

bdero commented on lib/ui/painting/image_decoder_impeller.cc:
```diff
@@ -370,32 +395,40 @@ ImageDecoderImpeller::UploadTextureToShared(
+            return;
+          }
+          blit_pass->SetLabel("Mipmap Blit Pass");
+          blit_pass->GenerateMipmap(texture);
```
The `texture.NeedsMipmapGeneration()` check also needs to be removed from `render_pass_mtl.mm`.

dnfield replied:
Should we change that to a DLOG on iOS?

dnfield replied:
I ifdef'd it out for IOS.

bdero replied:
I think we need to rethink these safety checks in general and have the HAL be less opinionated about it. While it's unsafe for GLES and probably worth tracking in some way, mipmaps can be uploaded directly (albeit, Impeller doesn't support this yet). One example is storing radiance maps for image based lighting.

bdero reviewed (approved):
LGTM
